bundles/org.openhab.binding.boschshc/src/main/resources/OH-INF/config/configs.xml

```diff
@@ -15,16 +16,34 @@
 			<description>The system password of the Bosch Smart Home Controller necessary for pairing.</description>
 		</parameter>
 	</config-description>
+
 	<config-description uri="thing-type:boschshc:device">
 		<parameter name="id" type="text" required="true">
 			<label>Device ID</label>
 			<description>Unique ID of the device.</description>
 		</parameter>
 	</config-description>
+
 	<config-description uri="thing-type:boschshc:user-defined-state">
 		<parameter name="id" type="text" required="true">
 			<label>State ID</label>
 			<description>Unique ID of the state.</description>
 		</parameter>
 	</config-description>
+
+	<config-description uri="thing-type:boschshc:device-with-two-child-devices">
+		<parameter name="id" type="text" required="true">
+			<label>Device ID</label>
+			<description>Unique ID of the device.</description>
+		</parameter>
+		<parameter name="childId1" type="text" required="true">
```

Comment:
@jlaur and @lsiepel I have a few questions here:

* Is it possible to introduce parameters that correspond to a `List<String>` here? It would be more generic if we had a parameter `childIds` instead of two parameters `childId1` and `childId2`.
* If this is not possible: is "camelCaseLowerCaseFirst" the right upper/lower case scheme? What about the word "ID"? Should I use `childId` or `childID`?
* The already introduced identifier for the main device ID is simply `id` (without "device"), so I aligned my names with `childId`. Or would you prefer to include `device` in the parameter name, like `childDeviceId1`?

Reply:
> @jlaur and @lsiepel I have a few questions here:
> 
> * Is it possible to introduce parameters that correspond to a `List<String>` here? It would be more generic if we had a parameter `childIds` instead of two parameters `childId1` and `childId2`.

Documentation is here: https://www.openhab.org/docs/developer/addons/config-xml.html#xml-structure-for-configuration-descriptions  There are bindings that use comma seperated list of values if that is what you seek. In the list of supported types, you can find the `default` and oir `multiple` that lists the comma seperated list. For reference: https://github.com/openhab/openhab-addons/pull/13988#issuecomment-1356814366 Do know that from data modeling i can understand this is preffered, but it might not be very user friendly to configure. 

> * If this is not possible: is "camelCaseLowerCaseFirst" the right upper/lower case scheme? What about the word "ID"? Should I use `childId` or `childID`?

According to: https://www.openhab.org/docs/developer/guidelines.html#b-code-formatting-rules-style the camalCase notation should be used for parameters, so yes `childId` is perfect.

> * The already introduced identifier for the main device ID is simply `id` (without "device"), so I aligned my names with `childId`. Or would you prefer to include `device` in the parameter name, like `childDeviceId1`?

Just `id` is fine, no need to add `device` to it.

Reply:
The devices and the corresponding IDs are contained in a list that can be requested from the Smart Home Controller. Here is a reduced version with the relevant devices (one parent device, two child devices):

```json
[{
    "@type": "device",
    "childDeviceIds": [
        "hdm:ZigBee:70ac08fffefead2d#3",
        "hdm:ZigBee:70ac08fffefead2d#2"
    ],
    "deviceModel": "MICROMODULE_LIGHT_CONTROL",
    "id": "hdm:ZigBee:70ac08fffefead2d",
    ...
}, {
    "@type": "device",
    "childDeviceIds": [],
    "deviceModel": "MICROMODULE_LIGHT_ATTACHED",
    "id": "hdm:ZigBee:70ac08fffefead2d#3",
    ...
}, {
    "@type": "device",
    "childDeviceIds": [],
    "deviceModel": "MICROMODULE_LIGHT_ATTACHED",
    "id": "hdm:ZigBee:70ac08fffefead2d#2",
    ...
}]
```

The child IDs can be found in two places:
* The `childDeviceIds` array in the parent device
* In the `id` properties of the child devices that are also contained in the top-level array

We currently see that the order of IDs in the `childDeviceIds` and the top-level array are consistent. However, the point is I am not sure whether the order (both in the `childDeviceIds` and in the top-level array) is deterministic and guaranteed to be always the same. Even now the order is not intuitive (`#3` before `#2`).

If this order is not stable we run into the unexpected and maybe even dangerous situation that we suddenly trigger another circuit than intended because they were interchanged. That's why I decided to "pin down" the child ID order via configuration.

Let me know whether you think this approach is viable, or whether I should maybe try to ask someone from Bosch whether the order is deterministic.

Reply:
You are right, since the child device IDs are not configurable/changeable by the user, configuration parameters might be the wrong approach. But it is not only possible to add the device automatically (through discovery), but also manually. Then the properties would not be set, is my assumption correct?

Anyway, the best solution would be to query the child IDs during the initialization of the parent device, then we would not even need properties. This would be possible if the order of the child device IDs is guaranteed to be the same every time. 

@sebastianharder or @cap-rb, can you help us here? Do you know whether the order of child device IDs (`childDeviceIds` of a Light Control II, see JSON example above) is stable, i.e. whether we get the child devices always in the same order?

Reply:
> Anyway, the best solution would be to query the child IDs during the initialization of the parent device, then we would not even need properties. This would be possible if the order of the child device IDs is guaranteed to be the same every time.

Exactly, I would try to complete 'hide' that it are childs for the user, so no configuration. If the device does not always garantee the same order, you can order it in the binding right after you parse the json, either way the order needs to be deterministic.

Reply:
> I just pushed a possible implementation (which includes sorting the child device IDs) and also replaced the JAR again 👍
> 
> @mike-bike can you please test again with this JAR? Please unlink the channels and delete the Light Control II thing before testing.

I will do this evening and report back.

> 
> I'm not convinced yet whether enforcing a certain child device order by sorting is the best way to go. @mike-bike, can you see a certain order of the two child devices (in your case `Brunnen` and `Olivenbaum`) in the Bosch app UI? Is there something layout-wise that indicates that one is the "first" and the other the "second" device/switch?

The order in the Bosch UI seems to be by default in alphabetical order of the Device names or probably in the sequence the devices have been added to the room. I had renamed "Brunnen" into "z Brunnen" and it moved last. Renaming back to "Brunnen" did not change the order again. But the tiles could be freely rearranged, so I conclude there is no fixed sequence in the UI which we should recognize... If it is only about the visual order, I wouldn't care less. 
If the channels will be stable after first initialization, we should be fine. I understood, that you need to have a stable sequence to achieve a deterministic mapping to the channels. If would always map the first logical device (say `#3`) to channel-1 and the second (`#2`) to channel-2 we would have an issue in case Bosch reports the devices in reverse sequence.

> 
> With the current implementation, the following will happen:
> 
> 1. We get the child IDs `hdm:ZigBee:70ac08fffefead2d#3` and `hdm:ZigBee:70ac08fffefead2d#2` from the SHC controller device info. `#3` corresponds to `Brunnen` and `#2` to `Olivenbaum`
> 2. We enforce a certain order by sorting the IDs (the order is now reversed because `#2` comes before `#3`)
> 3. The channel `power-switch-1` will now control `Olivenbaum`, and `power-switch-2` will control `Brunnen`
> 
> Please give us some feedback whether this is what you would expect as a user or not. Thanks for bearing with us and your testing efforts ❤️

The Bosch device has numbered outlets `1` and `2` for the wiring. My expectation would be to have channel-1 always on the wire `1` and channel-2 always on `2`. Unfortunately I can't recall the wiring of Brunnen/Olivenbaum and the switch is cramped into a box.

Though I do have a spare (still uninitialized) device. I am going to put some wires on, and document what Bosch is reporting.

Stay tuned